src/order/grade.lean

```diff
@@ -0,0 +1,334 @@
+/-
+Copyright (c) 2022 Yaël Dillies, Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Violeta Hernández Palacios, Grayson Burton, Vladimir Ivanov
+-/
+import data.nat.interval
+import data.nat.succ_pred
+import order.atoms
+
+/-!
+# Graded orders
+
+This file defines graded orders, also known as ranked orders.
+
+A graded order is an order with a bottom element in which every element has some finite "height",
+that corresponds to how many elements you need to get down by to reach `⊥`.
+
+## Main declarations
+
+* `grade_order`: Graded orders.
+* `grade`: The grade of an element.
+* `grade_max_order`: Graded orders with maximal elements. All maximal elements have the same grade.
+* `max_grade`: The maximum grade in a `grade_max_order`.
+* `order_embedding.grade`: The grade of an element in a linear order as an order embedding.
+
+## Implementation notes
+
+The standard approach is to define graded orders as the bounded orders whose flags (maximal chains)
+all have the same finite length (see Stanley p. 99).
+
+However, this means that graded orders all have a top and that we must reconstruct the grade of an
+element by how many elements we need to go down to reach `⊥`. Unpractical, really.
+
+Instead, we define graded orders by their grade function, without talking about flags yet.
+
+## References
+
+* [Konrad Engel, *Sperner Theory*][engel1997]
+* [Richard Stanley, *Enumerative Combinatorics*][stanley2012]
+-/
+
+open finset nat order_dual
+
+variables {α β : Type*}
+
+/-- A graded order is an order equipped with a grade function which tells how far a given element is
+away from the minimal elements. Precisely, `grade a` is the height of `a` in the Hasse diagram of
+`α`. -/
+class grade_order (α : Type*) [preorder α] :=
+(grade : α → ℕ)
+(grade_strict_mono : strict_mono grade)
+(grade_of_is_min ⦃a : α⦄ : is_min a → grade a = 0)
```

Comment:
Is this definition taken or adapted from any textbook? It would be nice to have a reference for this.

Reply:
Oh, I cooked up this definition using `is_min` myself, but upon rereading Engel, that's precisely his definition.

Reply:
Engel's definition has an additional requirement on the maximal elements of the preorder, so this is not the same thing.

Reply:
Yeah indeed it differs for the maximal elements. It's what he refers to as "the traditional definition" of a "ranked poset".

Reply:
Right, but do you have a reference showing this is actually used in some context? In particular it's not clear from the current references that this definition is mathematically useful.

Reply:
Conditions `grade`, `grade_strict_mono`, and `grade_of_covers` are all part of the standard definition. However, pretty much all graded posets I'm aware of in the wild (going off of the [Wikipedia example list](https://en.wikipedia.org/wiki/Graded_poset#Examples), for instance) have a single least element, so `grade_of_is_min` is strengthened to `grade_bot : grade ⊥ = 0`.

In my opinion, if we're not going to require `grade_bot` in this definition, `grade_of_is_min` is the sensible generalization, as it guarantees that the grade of an element is equal to the size of the largest chain it bounds. However, I doubt there will be any sources that back this decision for the reasons given prior.

Reply:
This comment is now outdated, but part of it still apply in the sense that our most general class (`grade_order`) seems to never be used when grading by `ℕ`, and instead only the next one will (`grade_min_order`). I believe this is fine, for the sake of completeness and because gradings by `ℤ` are sometimes encountered.